Dmr logic

dl1hrc opened the pull request:
Hi Tobias,
if you find the time please take a first look on my version of DmrLogic branch. It's in a very early state, at the moment the communication over the AMBEserver was tested. There are some points I would like to discuss.
73s de Adi / DL1HRC

sm0svx commented:
I'll see if I can find the time. I hope to do some work on DMR this summer but there are a lot of competing projects, some of them being non-computer related (!).

dl1hrc commented:
Ok, no problem. Remember that there is probably a lot of "bad" code inside that could be optimized.
It would help a lot if you only can take a look on the concept first, since we have changed the options-handling for the decoder/encoder classes to omit the setOptions() after creating the instances and implemented a multiton concept to create the DV dongle instances only once. Maybe you have different visions here. I will work on it in the near future...

sm0svx commented:
Never heard the term "multiton" before but it sounds like what I call "parameterized singleton", which is what I use in multiple places in SvxLink to encapsulate the shared use of unique resources (e.g. Async::SerialDevice and Async::AudioDevice).

sm0svx commented:
Note: I prefixed the description of this PR with "WIP" (Work in progress). That seem to be a standard way to differentiate between PRs that are ready to be merged and PRs that is there for collaboration reasons. When the branch is ready to be merged the "WIP" prefix can be removed.

porst17 commented:
Hi @sm0svx and @dl1hrc. I assisted @dl1hrc writing the AMBE related code and @dl1hrc asked me to add my two cents to this discussion (because it doesn't seem to move forward). The following is also related to #328.

IIRC the main problems with the hardware device for the codec were the following:
1. the hardware can not be initialized until all the necessary parameters (options) are known
1. the hardware is an encoder and a decoder at the same time
1. it can only be initialized once, so if you already created an encoder, you must not re-init the device if you create a decoder later
1. there are two ways to communicate with the hardware:
    1. direct connection via USB (USB-to-serial bridge?)
    1. indirect connection to a remote server via UDP which in turn communicates with the hardware directly

It was far from trivial to fulfill all of these requirements simultaneously. Here's how I solved it:
1. initialization: I switched from late option passing via `setOption()` to early option passing in the constructors of the `Audio..coder` (`..=En|De`). This follows the RAII principle more closely. Initialization errors in the contructors are fatal by design. If this happens, an exception can be thrown. I didn't think of graceful error handling to much at that time. This is subject to discussion.
1. one connection for de- and encoding: The class `AudioCodecAmbe` inherits both, `AudioEncoder` and `AudioDecoder`, i.e. it uses multiple inheritance. Usually, I try to avoid this, but in this case I saw no other practical option (besides refactoring even more of the existing code). Using a proxy object might be possible, but it would overcomplicate things IMO.
1. one time initialization: This uses the multion pattern: With a singleton, there is a unique instance of a class. The multiton allows to have multiple instances that are unique up to a certain key (in this case: the provided options).
1. several ways to access the device: The new class `AudioCodecAmbeDv3k` has multiple subclasses that can not be initialized directly, but that are accessed via a factory method that either returns a pre-existing multiton instance or creates (and stores) a new instance in case no previous instance with that configuration was found.

I implemented the [skeleton](https://github.com/porst17/svxlink/compare/550c96a2416141d75097a1d4aba78f8400adea3e...56862d4cb644df6047d3b460211fa36a202e570a) (containing the breaking changes) and @dl1hrc fleshed it out later without any significant contribution from my side.

I know that especially breaking changes are hard to integrate (and accept) into an existing project. But with the existing setup, it was very hard to fulfill the above requirements. I am open to discuss how to refactor the general design of the AMBE part of the `Audio..coder` subtree. We can either continue discussing it here or I can open a new pull request from [my old branch that contains only the skeleton](https://github.com/porst17/svxlink/tree/DmrLogic). This branch will never be merged, of course, but it is a lot easier to understand and would allow us to use GitHub's Review feature to directly discuss things right next to the code. Once the skeleton has a structure that everybody can live with, I can help @dl1hrc to refactor his code to match the refactored skeleton.

sm0svx commented:
Hi Christian and thanks for the details.

I was just about to ask about status for this PR but then @dl1hrc found a serious bug in the released code. I always try to prioritize serious bugs in the released code even though it's much more fun to work on new code but I guess I can do a little bit of both.

1. I think my comments in #328 is pretty clear. I do prefer to do initialization that can fail outside of the constructor because I have chosen not to use exceptions. The rest of SvxLink does not use exceptions so it would break coding style to start using it here. If I were already using exceptions, following RAII would be more logical, I agree. I don't see a problem with first constructing the object and then calling a new setOptions method. That should work in your case as well as in the existing code without requiring a lot of changes.
2. I do not shy away from multiple inheritance. I use it when I think it's appropriate but if it's not necessary I avoid it. I have never had any serious problem that have been caused by multiple inheritance.
3. Even though I have never called it "multiton" I have used the concept in SvxLink previously even though there is no specific class for it. I have called it "parameterized singleton". I have not looked at the code but I have no problem with this concept being used of course.
4. Yes. This challenge exist all over SvxLink. It's implemented in different ways and I have thought of  cleaning up the code but one example, that use the Factory.h class, is [src/trx/Ptt.h](/sm0svx/svxlink/blob/master/src/svxlink/trx/Ptt.h) being used in [LocalTx.cpp](/sm0svx/svxlink/blob/master/src/svxlink/trx/LocalTx.cpp#L295).

porst17 commented:
So it looks like your main concern is the `setOption()` thing. Here is the conflict:
- the `AudioCodecAmbeDv3k` is a [multiton](https://en.wikipedia.org/wiki/Multiton_pattern) because

    * there can only be one instance for each physical device/device configuration
    * the device is en- and decoder simultaniously and both modes are initialized at once without the ability to change parameters later on without starting from scratch.

  Hence, you need some kind of factory method to retrieve the en- and decoder object (which is in fact the same object) from different and probably unrelated parts of your code

- you want to first construct an object and later set its options. But then I can not utilize the multition pattern or anything that relies on a parameterized factory method to construct/retrieve the `Audio..coder` object. The factory needs the options and it needs them all at once because this the key of the map.

Without a multiton or at least something very similar, I have - at least right now - no idea how to satisfy the requirements from above. Maybe you have a better idea how to implement it. Please let me know. Any pseudo code is welcome.

The main problem is that you can not look at the problems in isolation. They are all connected, which made it particularly difficult to implement support for this device.

porst17 commented:
Just another question: Would it be a problem for you if we mix the old and the new approach? I my current implementation (I think) I just removed the public `setOption` method, because for the hardware device, you can not change codec options later on (IIRC you would need to reset and init the device again and this would affect the decoder if you are setting the option on the encoder and vice versa). We could add an intermediate class into the `Audio..coder` hierarchy that would add the public `setOption` method back again. So you could have codecs that support late changes to its options and codecs that don't.

Regarding the initialization of the DV3k device in the constructor: It is not strictly necessary to throw an exception here. The object could just initialize its internals with the given config and would try to connect to the device. It this fails, the constructor may still return without error and the object may try to reconnect later, just putting a note into the log file. This way we could also implement the graceful error handling you talked about in #328.

sm0svx commented:
My biggest concern with the codec stuff is that things that may fail (e.g. hardware initialization) is done in the constructor. Everything else I think I could accept, with the reservation that I have not gone through the code in detail.

Even if you handle device failures gracefully, as you suggest above, doing all initialization in the constructor make it harder to indicate if an error is fatal or not. You also force this more complicated behavior upon all other codecs.

As I see it, the main problem you face with the old codec framework is that it only supports selecting codec on type (e.g. GSM, SPEEX, AMBE etc) but you also need to construct (and share) new codec objects on device identity. What you have implemented does not only key on device identity but also on device configuration (e.g. TTY_BAUDRATE). It would be an impossible combination if two AMBE users opened the same device using different baudrates. The multiton would accept it though and happily create two objects that try to talk to the same device.

I'll give you some examples used in SvxLink where similar device sharing is used: For audio devices the multiton (or parameterized singleton) is keyd on the audio device name (e.g. alsa:plughw:0, oss:/dev/dsp, udp:127.0.0.1:1000 etc), for serial ports the key is the device node (e.g. /dev/ttyS0, /dev/ttyUSB0 etc), for PTY:s the path to the PTY is the key, for software defined receivers (DDR) a wideband receiver is shared between them and the key is a configuration section setting up the WBRX. I think you get the picture.

The problem of conflicting settings for the same device can be solved in different ways. For audio devices there is a global config setting up sample rate, channel count etc. For serial ports the last setting wins. For PTY:s there are no conflicting settings. For wideband receivers a device config section is used.

Maybe any of these solutions can solve the AMBE problem as well. We could extend the codec specification string to also accept a device ID, like in the audio device specification (e.g. AMBEUDP:127.0.0.1:10000, AMBETTY:/dev/xyz). What other parameters are required to change?  Maybe the baudrate isn't necessary to have selectable so it can be hard-coded? Are there any more parameters that have to be changeable?

I think it is wrong to treat device communications parameters as any other codec option. The codec options is there to specify codec parameters (like bitrate, quality, framesize etc). For example in the [NetUplink](https://github.com/sm0svx/svxlink/blob/master/src/svxlink/remotetrx/NetUplink.cpp#L615-L644) in RemoteTrx codec parameters are received from the client. Is it really logical that the remote station will decide which local hardware device that should be used? I'd say no. Is it logical for the remote station to decide other codec releated parameters? I'd say yes.

Another observation is that the user of an AMBE codec does not really care what device that is used since it's just a processing unit, so any unused device could be automatically assigned. This could be accomplished by implementing a AMBE codec device handler which handles all AMBE devices that are available in the system, both UDP and TTY variants or any other brand of codec. With "handler" I typically mean a singleton and not a separate application. Now it should be straight forward to implement a separate encoder and decoder that allocate a free AMBE codec from the AMBE handler. So now we're back at an essentially unmodified Encoder/Decoder framework where only the codec name need to be specified. Any device configuration is done elsewhere.

I kind of like the device handler solution. I'd probably end up adding an "initialize" method to the Decoder and Encoder that is called from the "AudioEn/Decoder::create" method. The initialize method could take the Options hash if that helps. If the initialization fail the object is destructed and NULL is returned from "create". The create method would also have to be extended to take the Options hash but that is not a problem. In the initialize method you can choose to return false if the AMBE device is unavailable. In an improved version one may do the device initialization more robust to temporary failures but it does not have to be there from the beginning.

The setOption function may not be necessary to keep as a public function if it feels awkward to handle it but it does not sound impossible. If it's just a question of resetting the device and reapplying existing settings at least it is possible even though it may cause a glitch in the audio stream for both coder and decoder. In a first implementation I would probably just ignore this a bit complicated case though.

If you want to start over with a new pull request that only contain the codec changes it's ok with me. That will make it easier to merge.

sm0svx commented:
Hmmm... That idea with AMBEUSB:devname does not work very well since we need to be able to request a codec from it's name only. This is the case with RemoteTrx and also in SvxReflector where the codec, at least at the moment, is set by the reflector server. Possibly the codec name could be modified locally before the request but it doesn't feel very transparent though.

So that brings us back to the "device handler/pool" approach. Maybe it's sufficient for a first implementation to only support one device if that is acceptable and reduce complexity?

sm0svx commented:
So we've got two main problems to solve now

* No option setting/hardware init in the constructor
* Don't mix device communication parameters with codec parameters. Must be able to choose a codec using name only.

porst17 commented:
Sorry for the long delay. I was quite busy the past two weeks and just now had the time to read your post in-depth.

Your analysis of the current situation is pretty much correct: 
- my key into the multiton map is wrong and includes too much. It mixes device identifier, device configuration and codec configuration.
- device configuration and codec parameters are two different things and could be done in separate places.

I also like the device handler approach. It seems pretty generic and may be utilized in other parts of svxlink later on. But unfortunately, I don't feel qualified to implement something like that into svxlink at the moment. I don't have any deep knowledge about the project and I most likely never will. I am not even a user, I am just assisting @dl1hrc when it comes to higher level software design and he somehow gets stuck.

So do you think it is possible that you could implement that at least partially? I am talking about the generic device handler only and how you imagine to access the device pool from different parts of svxlink, especially the codec part.

Regarding `setOption`: For our particular en-/decoder device, `setOption` is still problematic since it will cause the decoder to change its configuration even if you just called `setOption` on the encoder. I am not sure how to solve it. But the easiest option would be to have an init/create method the takes all codec parameters but no public `setOption` method. And certain subclasses could then provide a public `setOption` method if it makes sense for their codec. Also the current `setOption` that only changes one configuration value at a time may be problematic. Often, codec parameters are connected and you can not just change one (think of different encoding "modes"). So `setOption` should probably accept all configurable options at once.

So do you think it is feasible for you to create a basic implementation of the device handler (in a separate branch)? I and @dl1hrc would then try to refactor our code to fit onto the new framework. Would that be an option?

sm0svx commented:
I'm also interested in getting the support for this codec into SvxLink so I'll see what I can do this week-end. I'll probably have some time to work on it. The olympic hockey finals is not that interesting now or what do you think, @porst17 ;-)

I recognize the setOption problem and I guess that it could be possible to do away with it, maybe by setting all options in a new initialize method.

I'll get back to you when I have something working.

Could you (or @johnhays) help me get started by pointing me to some official documentation for the ThumbDV dongle?  I'll probably be able to reuse most of your code but it's always nice to have some documentation when proof-reading the code.

sm0svx commented:
Found the [datasheet for the DVSI-3000](http://www.dvsinc.com/manuals/AMBE-3000F_manual.pdf) chip.

sm0svx commented:
Status update: I've got a working test application talking to the DV3k that is built on a heavily modified version of the original AsyncAudioCodecAmbe code. I didn't get as far as to integrate it in the Coder/Decoder framework though. Hope to get working on that during next week.

porst17 commented:
Ok, great. That sounds like a good plan. I'll probably meet with @dl1hrc this Thursday, so if you have anything (preliminary) ready for us to check out, just let us know in advance!

sm0svx commented:
Mostly implemented in branch thumbdv.

- Only work for 8kHz audio at the moment.
- Need better error handling
- Serial protocol synchronization need to be done on startup
- /dev/ttyUSB0 hardcoded (no config reading yet)
- ...

sm0svx commented:
Support added for 16k sampling rate in branch thumbdv (actually only INTERNAL_SAMPLE_RATE=16000 is supported right now). Also created PR #376 for following up the AMBE codec stuff separately.
